File: fragments/api/gin/datastore/redis.go
package datastore

import (
	redis_crud "ff/redis_db"

	"github.com/gin-gonic/gin"
)

type FavouriteRequest struct {
	FavouriteCoin string `json:"favourite_coin" binding:"required"`
}

func RedisRoutes(r *gin.RouterGroup) {
	r.GET("/ping", RedisPing)
	favourites := r.Group("/favourites")

	favourites.GET("/:namespace", RedisGetFavourite)
	favourites.PUT("/:namespace", RedisCreateFavourite)
	favourites.PATCH("/:namespace", RedisUpdateFavourite)
	favourites.DELETE("/:namespace", RedisDeleteFavourite)
}

func RedisPing(c *gin.Context) {
	pong, err := redis_crud.RedisPing()
	if err != nil {
		c.JSON(500, gin.H{"error": "Internal Server Error"})
		return
	}
	c.JSON(200, gin.H{"message": pong})
}

func RedisGetFavourite(c *gin.Context) {
	namespace := c.Param("namespace")
	item, err := redis_crud.RedisRead(namespace)
	if err != nil {
		c.JSON(500, gin.H{"error": "Internal Server Error"})
		return
	}
	if item["favourite_coin"] == "" {
		c.Status(404)
		return
	}
	c.JSON(200, item)
}

func RedisCreateFavourite(c *gin.Context) {
	namespace := c.Param("namespace")
	var req FavouriteRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(422, gin.H{"error": "Invalid payload"})
		return
	}
	_, err := redis_crud.RedisCreate(namespace, req.FavouriteCoin)
	if err != nil {
		c.JSON(500, gin.H{"error": "Internal Server Error"})
		return
	}
	c.Status(200)
}

func RedisUpdateFavourite(c *gin.Context) {
	namespace := c.Param("namespace")
	var req FavouriteRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(422, gin.H{"error": "Invalid payload"})
		return
	}
	_, err := redis_crud.RedisUpdate(namespace, req.FavouriteCoin)
	if err != nil {
		c.JSON(500, gin.H{"error": "Internal Server Error"})
		return
	}
	c.Status(200)
}

func RedisDeleteFavourite(c *gin.Context) {
	namespace := c.Param("namespace")
	_, err := redis_crud.RedisDelete(namespace)
	if err != nil {
		c.JSON(500, gin.H{"error": "Internal Server Error"})
		return
	}
	c.Status(204)
}

File: fragments/apis/fastify/datastore/redis.ts
import { type FastifyInstance, type FastifyPluginOptions } from "fastify";
import { redisCreate, redisDelete, redisPing, redisRead, redisUpdate } from "../../../redis_db/redis_crud";

export const routes = (fastify: FastifyInstance, _: FastifyPluginOptions) => {
  fastify.get<{
    Reply: {
      200: { message: string };
      500: { error: string };
    };
  }>("/ping", async (request, reply) => {
    try {
      const result = await redisPing();
      return { message: result };
    } catch (error) {
      request.log.error(error, "Error pinging Redis");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.get<{
    Params: { namespace: string };
    Reply: {
      200: { favourite_coin?: string };
      404: void;
      500: { error: string };
    };
  }>("/favourites/:namespace", async (request, reply) => {
    try {
      const { namespace } = request.params;
      const item = await redisRead(namespace);

      if (!item.favourite_coin) {
        return reply.code(404).send();
      }
      return item;
    } catch (error) {
      request.log.error(error, "Error reading favourite coin from Redis");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.put<{
    Body: { favourite_coin: string };
    Params: { namespace: string };
    Reply: {
      200: void;
      400: { error: string };
      500: { error: string };
    };
  }>("/favourites/:namespace", {
    schema: {
      body: {
        type: "object",
        required: ["favourite_coin"],
        properties: {
          favourite_coin: { type: "string" },
        },
      },
    },
  }, async (request, reply) => {
    try {
      const { namespace } = request.params;
      const { favourite_coin } = request.body;
      const result = await redisCreate(namespace, favourite_coin);
      if (result !== "OK") {
        throw new Error(result);
      }
      return reply.code(200).send();
    } catch (error) {
      request.log.error(error, "Error creating favourite coin in Redis");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.patch<{
    Body: { favourite_coin: string };
    Params: { namespace: string };
    Reply: {
      200: void;
      500: { error: string };
    };
  }>("/favourites/:namespace", {
    schema: {
      body: {
        type: "object",
        required: ["favourite_coin"],
        properties: {
          favourite_coin: { type: "string" },
        },
      },
    },
  }, async (request, reply) => {
    try {
      const { namespace } = request.params;
      const { favourite_coin } = request.body;
      const result = await redisUpdate(namespace, favourite_coin);
      if (result !== "OK") {
        throw new Error(result);
      }
      return reply.code(200).send();
    } catch (error) {
      request.log.error(error, "Error updating favourite coin in Redis");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });

  fastify.delete<{
    Params: { namespace: string };
    Reply: {
      204: void;
      500: { error: string };
    };
  }>("/favourites/:namespace", async (request, reply) => {
    try {
      const { namespace } = request.params;
      const result = await redisDelete(namespace);
      if (result !== "OK") {
        throw new Error(result);
      }
      return reply.code(204).send();
    } catch (error) {
      request.log.error(error, "Error deleting favourite coin from Redis");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });
};
